### PR TITLE
Version bump to etcd v3.2.11, grpc v1.7.5

### PR DESCRIPTION
**Pull request by jpbetz**
Fix https://github.com/kubernetes/kubernetes/issues/56114: Update to etcd client 3.2.11

Version bumps:

- etcd from 3.1.10 to 3.2.11
- grpc from 1.3.0 to 1.7.5
- grpc-gateway from v1.1.0-25-g84398b9 to v1.3.0

TODO:

- [x] Apply etcd [3.2 client upgrade guide](https://github.com/coreos/etcd/blob/master/Documentation/upgrades/upgrade_3_2.md)
- [x] Apply grpc API changes in 1.6.0 and 1.7.0 [release notes](https://github.com/grpc/grpc-go/releases)
- [x] bbolt was pulled in transitively, why? We have tests that embed etcd, so we must vendor the etcd server and all it's dependencies.
- [x] Upgrade to containerd v1.0.0? Currently kubernetes depends on containerd v1.0.0-beta.2-159-g27d450a0 which depends on grpc v1.3.0, but  containerd v1.0.0 depends on grpc 1.7.2. Not needed. The containerd grpc upgrade required [no code changes](https://github.com/containerd/containerd/commit/ce3e32680d4d27fb94277509a49f3ce4ef5b05a8).
- [x] Fix all failing tests
- [x] Ensure we can safely upgrade grpc to 1.7.5 given that docker and cAdvisor still depend on grpc 1.3.0 (both in the versions we vend and on master for both projects). Should we hold off on this change until we have a docker release that uses gprc 1.7.x?
- [x] Wait for grpc 1.7.5 to be released (it will include https://github.com/grpc/grpc-go/pull/1747). Once released, bump grpc version in this PR and remove workarounds in `hack/godep-save.sh`.

Transitive dependencies on grpc:
- docker depends on grpc, but according to the package dependency graph (`go list -f '{{ .Deps }}'`) there are no dependencies from kubernetes to grpc via docker packages.
- containerd v1.0.0 depends on grpc 1.7.2, we should upgrade to containerd v1.0.0 soon, this can be done in a separate PR
- cadvisor depends on grpc 1.3.0 on master, it should upgrade it to grpc 1.7.5, this can be done in a separate PR

**Release note**:

```release-note
Upgrade to etcd client 3.2.11 and grpc 1.7.5 to improve HA etcd cluster stability.
```

### CODE REVIEW

**Comment by luxas** [comment]
cc @kubernetes/sig-cluster-lifecycle-pr-reviews @kubernetes/sig-scalability-pr-reviews @kubernetes/sig-api-machinery-pr-reviews

**Comment by jpbetz** [comment]
This minimally reproduces the godep issue `pull-kubernetes-verify` is having:

https://gist.github.com/jpbetz/2909720a2276a0303ca68f92b776a7c1

**Comment by jpbetz** [comment]
I've filed an issue with godep: https://github.com/tools/godep/issues/554

Putting a workaround in place now

**Comment by thockin** [comment]
Vendor stuff LGTM as much as I can verify.  Who can review the rest of this?

File: hack/godep-save.sh [hunk 1]
```diff
@@ -62,12 +62,25 @@ REQUIRED_BINS=(
   "./..."
 )
 
+# Workaround for https://github.com/tools/godep/issues/554 until: (a) it is resolved, (b) grpc adds a doc.go
```
Comment (jpbetz) [inline hack/godep-save.sh]: Not planning to merge this. It's in place to see if https://github.com/tools/godep/issues/554 is the only dependency issue in the e2e tests.

File: hack/godep-save.sh [hunk 2]
```diff
@@ -62,12 +62,25 @@ REQUIRED_BINS=(
...
+  echo "package grpc_lb_v1" > "${GOPATH}/src/google.golang.org/grpc/grpclb/grpc_lb_v1/doc.go"
+fi
+
 kube::log::status "Running godep save - this might take a while"
 # This uses $(pwd) rather than ${KUBE_ROOT} because KUBE_ROOT will be
 # realpath'ed, and godep barfs ("... is not using a known version control
 # system") on our staging dirs.
 GOPATH="${GOPATH}:$(pwd)/staging" godep save "${REQUIRED_BINS[@]}"
 
+# Workaround for https://github.com/tools/godep/issues/554: strip empty grpc_lb_v1 package from Godeps.
```
Comment (jpbetz) [inline hack/godep-save.sh]: Same here, not planning to merge. Just testing.

**Comment by jpbetz** [comment]
@gyuho Would you have a quick look at the etcd related changes when you get a sec? There are only a couple lines.

**Comment by gyuho** [comment]
@jpbetz Sure, I will take a look by tomorrow.

FYI, we've just found another regression in etcd server-side. The fix https://github.com/coreos/etcd/pull/9018 will be merged tomorrow, with another v3.2 release planned next week. It would be great if we can also get https://github.com/grpc/grpc-go/pull/1743 in the new gRPC v1.7.5 release and in the new etcd v3.2.12 release.

**Comment by gyuho** [comment]
@jpbetz etcd changes lgtm. Thanks!

**Review by timothysc — COMMENTED** [review]
/lgtm - might want to have @wojtek-t take a look.  I'll be out until the new year.

**Comment by wojtek-t** [comment]
LGTM too - @jpbetz please ping me when this will be ready for merging.

**Comment by jpbetz** [comment]
@wojtek-t ready to merge.  I just updated to latest grpc to fix a godep issue. Would you lgtm this when you're ready?

**Comment by ash2k** [comment]
FYI this fixes https://github.com/kubernetes/kubernetes/issues/53124. Previous attempt to bump gRPC to 1.6.0 (https://github.com/kubernetes/kubernetes/pull/53128) was reverted (https://github.com/kubernetes/kubernetes/pull/54400) because it caused scalability issues - see https://github.com/kubernetes/kubernetes/issues/51099. I wonder if the issue was solved or there is a workaround in place.

**Comment by wojtek-t** [comment]
> FYI this fixes #53124. Previous attempt to bump gRPC to 1.6.0 (#53128) was reverted (#54400) because it caused scalability issues - see #51099. I wonder if the issue was solved or there is a workaround in place.

@porridge @shyamjvs @kubernetes/sig-scalability-misc ^^

**Comment by porridge** [comment]
> > FYI this fixes #53124. Previous attempt to bump gRPC to 1.6.0 (#53128) was reverted (#54400) because it caused scalability issues - see #51099. I wonder if the issue was solved or there is a workaround in place.
> 
> @porridge @shyamjvs @kubernetes/sig-scalability-misc ^^

Were there any tests performed to ensure this does not cause scalability issues again?

**Comment by porridge** [comment]
I can answer my own question: it was most likely not tested, because it broke scalability tests again, in almost exactly the same way as last time: https://github.com/kubernetes/kubernetes/issues/51099#issuecomment-352765584
Please revert this back ASAP, and only roll forward after it has been shown to pass at large scale.

**Comment by wojtek-t** [comment]
Revert is here: https://github.com/kubernetes/kubernetes/pull/57407

**Comment by xiang90** [comment]
@porridge 

etcd server now sets the max response size to unlimited. we back ported it to 3.2. https://github.com/coreos/etcd/pull/8867/commits/939337f45034585ee5203a4a83429fc45a15ddb8. 

not sure if it is the exact same problem.

**Comment by gyuho** [comment]
@xiang90 Seems like etcd needs set larger call size in client-side? https://github.com/grpc/grpc-go/blob/2720857d978b9873027f6d739d74794428f49c03/rpc_util.go#L214-L228

if the error message in https://github.com/kubernetes/kubernetes/issues/51099 is from etcd clientv3 calls

**Comment by jpbetz** [comment]
@gyuho based on the [stack trace from apiserver logs](https://gist.github.com/jpbetz/f5f597d53cd06c5c77eca6882acb2c65), it looks like the error originated at this [etcd get call](https://github.com/kubernetes/kubernetes/blob/6b03a43b761a8b6f9eac3648cf499f157f161414/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go#L556) in the kubernetes source code and was logged by this [error handler](https://github.com/kubernetes/kubernetes/blob/56acf19e6d15c74280765d76de4fb8e62b570924/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/status.go#L64).

**Comment by gyuho** [comment]
@jpbetz I see.

Then, upstream etcd clientv3 needs to bump up the recv call size limit to its gRPC calls (https://github.com/coreos/etcd/issues/9043), since gRPC defaults its limit to 4MB when not configured.

**Comment by jpbetz** [comment]
Thanks @gyuho!

I'll shift over to working on:
- making sure this is handled in the kubernetes go client
- creating a simple end-to-end test to reproduce this issue both at the etcd and kubernetes go client levels so we can quickly verify the fixes resolve the issue as they become available

**Comment by jpbetz** [comment]
This is superseded by https://github.com/kubernetes/kubernetes/pull/57480, which will bump the etcd version to 3.2.12

**Pull request merged**